### PR TITLE
Prevent possible PHP warnings produced by 'woocommerce_ajax_order_items_removed' hook

### LINKED ISSUE #32717: PHP Warning: Undefined variable $changed_stock" in includes/class-wc-ajax.php on line 1336
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Seen in the log today:

>PHP Warning:  Undefined variable $changed_stock in /srv/wp-content/plugins/woocommerce/includes/[class-wc-ajax.php on line 1336](https://github.com/woocommerce/woocommerce/blob/6.4.1/plugins/woocommerce/includes/class-wc-ajax.php#L1336)

Unfortunately I don't know what caused it, but looking at the affected line & above, that variable is [indeed only defined](https://github.com/woocommerce/woocommerce/blob/6.4.1/plugins/woocommerce/includes/class-wc-ajax.php#L1307) after two `if` checks – rather than being defined ahead or covered by an `else` branch.

I take it's the second check that caused the problem here as otherwise the first two parameters `$item_id` & `$item` would have been undefined as well. Seems those two could use a default, too.

### Expected behavior

Nobody complained, so seems harmless.

### Actual behavior

No entry in the debug log.

### Steps to reproduce

1. Use WooCommerce
2. Don't know, unfortunately

### WordPress Environment

`
### WordPress Environment ###

WC Version: 6.4.1
REST API Version: ✔ 6.4.1
WC Blocks Version: ✔ 7.2.2
Action Scheduler Version: ✔ 3.4.0
WC Admin Version: ✔ 3.3.2
Log Directory Writable: ✔
WP Version: 5.9.3
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache
PHP Version: 8.0.17
PHP Post Max Size: 120 MB
PHP Time Limit: 0
PHP Max Input Vars: 5000
cURL Version: 7.61.1
OpenSSL/1.1.1k

SUHOSIN Installed: –
MySQL Version: 10.4.24-MariaDB
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 6.4.1
WC Database Prefix: wp_

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (50) ###

Complianz | GDPR/CCPA Cookie Consent: by Really Simple Plugins – 6.1.5
DHL for WooCommerce: by DHL – 2.8.12
WooCommerce Admin: by WooCommerce – 3.3.2
German Market: by MarketPress – 3.14.1
WooCommerce: by Automattic – 6.4.1
Yoast SEO: by Team Yoast – 18.6

### Dropin Plugins (1) ###

advanced-cache.php: advanced-cache.php

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: right_space
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
gift-card (gift-card)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #4390616 - /neu-im-sortiment/
Cart: #20 - /warenkorb/
Checkout: #21 - /kasse/
My account: #22 - /mein-konto/
Terms and conditions: #40 - /agb/

### Theme ###

Name: Divi Jewellery Shop
Version: 1.0.1
Author URL: http://divi-childthemes.com
Child Theme: ✔
Parent Theme Name: Divi
Parent Theme Version: 4.17.1
Parent Theme Author URL: http://www.elegantthemes.com
WooCommerce Support: ✔

### Action Scheduler ###

Complete: 4,269
Oldest: 2022-04-02 10:22:28 +0200
Newest: 2022-04-21 14:29:49 +0200

Pending: 6
Oldest: 2022-04-21 15:29:49 +0200
Newest: 2022-04-28 12:37:51 +0200

### Status report information ###

Generated at: 2022-04-21 14:31:24 +02:00
`

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [ ] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by jorgeatorres**
### All Submissions:

-   [X] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [X] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Under certain circumstances, when removing an order item via admin, some PHP warnings could be triggered due some possibly undefined variables [when firing hook `woocommerce_ajax_order_items_removed`](https://github.com/woocommerce/woocommerce/blob/5019a344a1bb762c5bccd54d989f17b25dfb8613/plugins/woocommerce/includes/class-wc-ajax.php#L1346-L1356).

The hook receives arguments that are only available inside a [`foreach` loop above](https://github.com/woocommerce/woocommerce/blob/5019a344a1bb762c5bccd54d989f17b25dfb8613/plugins/woocommerce/includes/class-wc-ajax.php#L1321-L1339) so that's where the `do_action()` call should probably live. Even when those vars are set, the hook would only fire with the last value of those vars, which seems incorrect.

This PR moves the hook to the correct location. Given that, under normal circumstances, only one order item is passed via the admin to this AJAX endpoint, the behavior of only firing once should be preserved.

Closes #32717.

### How to test the changes in this Pull Request:

- [X] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

While our current code is certainly prone to generating the PHP warnings mentioned, under normal usage, those shouldn't be triggered, so maybe users are encountering this through other plugins/etc. Because of this the testing instructions are a bit artificial.
In any case, plugging the hole in core is probably the way to go.

1. Go to WC > Orders.
2. Create a test order with a few products.
3. Open the browser's JS console and execute the following command:
   ```js
   jQuery.ajax({url: woocommerce_admin_meta_boxes.ajax_url, data: { order_id: woocommerce_admin_meta_boxes.post_id, order_item_ids: [1], action: 'woocommerce_remove_order_item', security: woocommerce_admin_meta_boxes.order_item_nonce }, type: 'POST' });  
   ```
4.
   - On `trunk`, you should see a few PHP warnings like these:
      > PHP Warning:  foreach() argument must be of type array|object, string given in [...]woocommerce/includes/class-wc-ajax.php on line 1321
      > PHP Warning:  Undefined variable $item_id in [...]woocommerce/includes/class-wc-ajax.php on line 1356
      > PHP Warning:  Undefined variable $item in [...]woocommerce/includes/class-wc-ajax.php on line 1356
      > PHP Warning:  Undefined variable $changed_stock in [...]woocommerce/includes/class-wc-ajax.php on line 1356      
   - On this branch, no warnings should be logged.
5. Confirm that you are able to add and remove products and order items from the order successfully.

<!-- End testing instructions —>

### Other information:

-   [X] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [X] Have you written new tests for your changes, as applicable?
-   [X] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [X] Have you included testing instructions?

<!-- Mark completed items with an [x] —>

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Review by coreymckrill — COMMENTED** [review]
🤔 I'm concerned that there will be a side effect of moving the action hook up into the foreach loop. It seems like the 4th parameter, `$order`, would potentially have a different value than it does in the original location, since after the foreach loop the order gets updated and the tax and totals get recalculated.

It almost seems like you need a hook in both places, but the one in the original location should have different parameters, like an array of order item IDs, instead of just the last one.

**Comment by jorgeatorres** [comment]
> It seems like the 4th parameter, `$order`, would potentially have a different value than it does in the original location, since after the foreach loop the order gets updated and the tax and totals get recalculated.

You're right. After thinking about it, I believe we should leave the hook where it is but check that the vars are set and provide a default when not (as to suppress the possible PHP warnings).

That way the previous behavior ("incorrect" or not) is preserved, and given that at most one order item is (normally) passed to this endpoint, it wouldn't make a huge difference, except that `$order` would be up-to-date.

> 
> It almost seems like you need a hook in both places, but the one in the original location should have different parameters, like an array of order item IDs, instead of just the last one.

Yes. I agree. I'm pretty certain the hook was added to its current location by mistake (as the args don't make much sense in its current context) but unless someone needs it, adding a new hook is probably something we shouldn't do right now.

- - - 

If my suggestion of moving the hook back to its original place but add a check for the vars makes sense to you, I'll do that tomorrow. Let me know @coreymckrill.

**Comment by coreymckrill** [comment]
@jorgeatorres 

> If my suggestion of moving the hook back to its original place but add a check for the vars makes sense to you, I'll do that tomorrow.

Yeah that sounds good. It seems like there is something we could do here with deprecating the existing hook and replacing it with one that makes more sense, but it probably isn't worth the hassle at this point.

**Comment by jorgeatorres** [comment]
Hey @coreymckrill!

Can please you take a look again? I agree that we should probably replace this hook but it feels like it's not a very popular one (only 1 issue about it) so maybe this will do for now: prevents the PHP warning and doesn't really change behavior.

**Review by coreymckrill — APPROVED** [review]
👍 The test works, this seems like a good way to resolve the immediate issue.
